File: hs/codewars/sort_binary_tree_by_levels.hs
-- https://www.codewars.com/kata/sort-binary-tree-by-levels

module TreeByLevels where

import TreeByLevels.TreeNode

treeByLevels :: Maybe (TreeNode a) -> [a]
treeByLevels x = do
  nodes <- levels x
  map value nodes

levels :: Maybe (TreeNode a) -> [[(TreeNode a)]]
levels Nothing = []
levels (Just node) = ([node]:others) where
  others = zipWithDefault [] [] (++) (levels $ left node) (levels $ right node)

zipWithDefault :: a -> b -> (a -> b -> c) -> [a] -> [b] -> [c]
zipWithDefault dx _  f []     ys     = zipWith f (repeat dx) ys
zipWithDefault _  dy f xs     []     = zipWith f xs (repeat dy)
zipWithDefault dx dy f (x:xs) (y:ys) = f x y : zipWithDefault dx dy f xs ys

File: py/codewars/sort_binary_tree_by_levels.py
#!/usr/bin/env python3
# https://www.codewars.com/kata/sort-binary-tree-by-levels

# Iterative solution


def tree_by_levels(node):
    if node is None:
        return []

    level = [node]
    levels = [[node.value]]

    while len(level) != 0:
        level = [y for x in level for y in (x.left, x.right) if y is not None]
        levels.append([x.value for x in level])

    return [x for xs in levels for x in xs]


# Recursive solution

from itertools import chain, zip_longest


def levels(node):
    if node is None:
        return
    yield [node]
    yield from (
        chain(l, r)
        for l, r in zip_longest(
            levels(node.left), levels(node.right), fillvalue=[]
        )
    )


def tree_by_levels(node):
    return [y.value for x in levels(node) for y in x]
